File: WordPress/Classes/Models/MenuItem.m
#import "MenuItem.h"
#import "Menu.h"
#import "PostType.h"
#import "Blog.h"

NSString * const MenuItemTypePage = @"page";
NSString * const MenuItemTypeCustom = @"custom";
NSString * const MenuItemTypeCategory = @"category";
NSString * const MenuItemTypeTag = @"post_tag";
NSString * const MenuItemTypePost = @"post";
NSString * const MenuItemTypeJetpackTestimonial = @"jetpack-testimonial";
NSString * const MenuItemTypeJetpackPortfolio = @"jetpack-portfolio";
NSString * const MenuItemTypeJetpackComic = @"jetpack-comic";

NSString * const MenuItemLinkTargetBlank = @"_blank";
NSString * const MenuItemDefaultLinkTitle = @"New Item";

@implementation MenuItem

@dynamic contentID;
@dynamic details;
@dynamic itemID;
@dynamic linkTarget;
@dynamic linkTitle;
@dynamic name;
@dynamic type;
@dynamic typeFamily;
@dynamic typeLabel;
@dynamic urlStr;
@dynamic menu;
@dynamic children;
@dynamic parent;
@dynamic classes;

+ (NSString *)entityName
{
    return NSStringFromClass([self class]);
}

+ (NSString *)labelForType:(NSString *)itemType blog:(nullable Blog *)blog
{
    NSString *label = nil;
    if ([itemType isEqualToString:MenuItemTypePage]) {
        label = NSLocalizedString(@"Page", @"Menu item label for linking a page.");
    } else if ([itemType isEqualToString:MenuItemTypePost]) {
        label = NSLocalizedString(@"Post", @"Menu item label for linking a post.");
    } else if ([itemType isEqualToString:MenuItemTypeCustom]) {
        label = NSLocalizedString(@"Link", @"Menu item label for linking a custom source URL.");
    } else if ([itemType isEqualToString:MenuItemTypeCategory]) {
        label = NSLocalizedString(@"Category", @"Menu item label for linking a specific category.");
    } else if ([itemType isEqualToString:MenuItemTypeTag]) {
        label = NSLocalizedString(@"Tag", @"Menu item label for linking a specific tag.");
    } else if ([itemType isEqualToString:MenuItemTypeJetpackTestimonial]) {
        label = NSLocalizedString(@"Testimonials", @"Menu item label for linking a testimonial post.");
    } else if ([itemType isEqualToString:MenuItemTypeJetpackPortfolio]) {
        label = NSLocalizedString(@"Projects", @"Menu item label for linking a project page.");
    } else if ([itemType isEqualToString:MenuItemTypeJetpackComic]) {
        label = NSLocalizedString(@"Comics", @"Menu item label for linking a comic page.");
    } else if (blog) {
        // Check any custom postTypes that may have a label for the itemType.
        for (PostType *postType in blog.postTypes) {
            // If the postType name matches, use its label.
            if ([postType.name isEqualToString:itemType]) {
                label = postType.label;
                break;
            }
        }
    }
    return label;
}

+ (NSString *)defaultItemNameLocalized
{
    return NSLocalizedString(@"New item", @"Menu item title text used as default when creating a new menu item.");
}

/**
 Traverse parent's of the item until we reach nil or a parent object equal to self.
*/
- (BOOL)isDescendantOfItem:(MenuItem *)item
{
    BOOL otherItemIsDescendant = NO;
    MenuItem *ancestor = self.parent;
    while (ancestor) {
        if (ancestor == item) {
            otherItemIsDescendant = YES;
            break;
        }
        ancestor = ancestor.parent;
    };
    return otherItemIsDescendant;
}

/**
 Traverse the orderedItems for parent items equal to self or that are a descendant of self (a child of a child).
 */
- (MenuItem *)lastDescendantInOrderedItems:(NSOrderedSet *)orderedItems
{
    MenuItem *lastChildItem = nil;
    NSUInteger parentIndex = [orderedItems indexOfObject:self];
    for (NSUInteger i = parentIndex + 1; i < orderedItems.count; i++) {
        MenuItem *child = [orderedItems objectAtIndex:i];
        if (child.parent == self) {
            lastChildItem = child;
        }
        if (![lastChildItem isDescendantOfItem:self]) {
            break;
        }
    }
    return lastChildItem;
}

- (BOOL)nameIsEmptyOrDefault
{
    return self.name.length == 0 || [self.name isEqualToString:[MenuItem defaultItemNameLocalized]];
}

@end

File: Sources/WordPressData/Swift/MenuItem.swift
import Foundation
import CoreData

@objc(MenuItem)
public class MenuItem: NSManagedObject {

    @objc public static let linkTargetBlank = "_blank"
    @objc public static let defaultLinkTitle = "New Item"

    @objc(labelForType:blog:)
    public static func label(for itemType: String?, blog: Blog? = nil) -> String? {
        guard let itemType else { return nil }

        switch itemType {
        case MenuItemType.page:
            return NSLocalizedString(
                "Page",
                comment: "Menu item label for linking a page."
            )
        case MenuItemType.post:
            return NSLocalizedString(
                "Post",
                comment: "Menu item label for linking a post."
            )
        case MenuItemType.custom:
            return NSLocalizedString(
                "Link",
                comment: "Menu item label for linking a custom source URL."
            )
        case MenuItemType.category:
            return NSLocalizedString(
                "Category",
                comment: "Menu item label for linking a specific category."
            )
        case MenuItemType.tag:
            return NSLocalizedString(
                "Tag",
                comment: "Menu item label for linking a specific tag."
            )
        case MenuItemType.jetpackTestimonial:
            return NSLocalizedString(
                "Testimonials",
                comment: "Menu item label for linking a testimonial post."
            )
        case MenuItemType.jetpackPortfolio:
            return NSLocalizedString(
                "Projects",
                comment: "Menu item label for linking a project page."
            )
        case MenuItemType.jetpackComic:
            return NSLocalizedString(
                "Comics",
                comment: "Menu item label for linking a comic page."
            )
        default:
            for postType in (blog?.postTypes ?? []) {
                if let postType = postType as? PostType, postType.name == itemType {
                    return postType.label
                }
            }
            return nil
        }
    }

    @objc public static func defaultItemNameLocalized() -> String {
        return NSLocalizedString(
            "New item",
            comment: "Menu item title text used as default when creating a new menu item."
        )
    }

    @objc(isDescendantOfItem:)
    public func isDescendant(of item: MenuItem) -> Bool {
        var ancestor = parent
        while let current = ancestor {
            if current == item {
                return true
            }
            ancestor = current.parent
        }
        return false
    }

    @objc(lastDescendantInOrderedItems:)
    public func lastDescendant(in orderedItems: NSOrderedSet) -> MenuItem? {
        let parentIndex = orderedItems.index(of: self)
        guard parentIndex != NSNotFound else { return nil }

        var lastChildItem: MenuItem? = nil
        for i in (parentIndex + 1)..<orderedItems.count {
            if let child = orderedItems.object(at: i) as? MenuItem {
                if child.parent == self {
                    lastChildItem = child
                }
                if let lastChildItem, !lastChildItem.isDescendant(of: self) {
                    break
                }
            }
        }
        return lastChildItem
    }

    @objc
    public var nameIsEmptyOrDefault: Bool {
        guard let name else { return true }
        return name.isEmpty || name == MenuItem.defaultItemNameLocalized()
    }

    @objc(precedingSiblingInOrderedItems:)
    public func precedingSibling(in orderedItems: NSOrderedSet) -> MenuItem? {
        let selfIndex = orderedItems.index(of: self)
        guard selfIndex != NSNotFound else { return nil }

        for idx in stride(from: selfIndex - 1, through: 0, by: -1) {
            if let previousItem = orderedItems.object(at: idx) as? MenuItem {
                if previousItem.parent == parent {
                    return previousItem
                }
            }
        }
        return nil
    }
}

// MARK: - Constants

@objcMembers public class MenuItemType: NSObject {
    public static let page = "page"
    public static let custom = "custom"
    public static let category = "category"
    public static let tag = "post_tag"
    public static let post = "post"
    public static let jetpackTestimonial = "jetpack-testimonial"
    public static let jetpackPortfolio = "jetpack-portfolio"
    public static let jetpackComic = "jetpack-comic"
}
